### ISSUE TITLE
Compiler segfaulting on linux

### ISSUE THREAD

**Issue by LeoBlute** [odin-lang/Odin#4224]
The compiler will randomly throw a segfault instead of generating the binary normally, the problem is related with the llvm section, taking main() as the starting point, the crash occurs when lb_generate_code() is called, going a little further, i've tracked the bug to, at least, the function lb_setup_type_info_data_giant_array() at: https://github.com/odin-lang/Odin/blob/master/src/llvm_backend_type.cpp#L1124
It's not the final source of the segfault, so it should be investigated further
Undefined Behavior is likely the cause, as the crash will happen sometimes only, while other will compile the code properly

**Comment by Kelimion** [comment]
It would be nice to investigate, but you've left us next to nothing to go on. Odin successfully runs and compiles code on Linux all the time.

At the very least we'll need the output of `odin report` so we know the version of Odin, *what* Linux, which which version of LLVM it was compiled against.

And preferably also a minimum reproductive code example that will result in this segfault, even if it's only once every 10 compiles.

**Comment by Kelimion** [comment]
Does it still occur with `-no-threaded-checker`?

**Comment by harold-b** [comment]
FWIW, Jeroen, I started getting random segfaults as well on macOS:

```
        Odin:    dev-2024-09:71a8de035
        OS:      macOS Sonoma 14.5 (build: 23F79, kernel: 23.5.0)
        CPU:     Apple M3 Max
        RAM:     65536 MiB
        Backend: LLVM 14.0.6
        ```

I haven't investigated yet, but it started triggering fairly often. I'll try to get more info tonight after work, or tomorrow, to see if I can provide any useful information.

**Comment by harold-b** [comment]
Opened #4228 in case it is unrelated.

**Comment by LeoBlute** [comment]
Sorry for not replying earlier, here is the report:
        Odin:    dev-2024-09:90358f97d
        OS:      Ubuntu 22.04.4 LTS, Linux 6.8.8-060808-generic
        CPU:     AMD Ryzen 3 2200G with Radeon Vega Graphics    
        RAM:     13889 MiB
        Backend: LLVM 14.0.0

Using -no-threaded-checker, actually, makes the crash occur more ofter
I'm going to isolate my code to find what part of it makes this crash happen, and then i'll send it

**Comment by laytan** [comment]
@harold-b 's report was also llvm 14 so it may be isolated to that version. Also, thanks for the detailed report!

**Comment by LeoBlute** [comment]
> @LeoBlute would you be able to test on the latest nightly release? That statically links LLVM18 and i can't reproduce the segfault on my Fedora machine (which uses LLVM18)

Tested, this segfault does not occurs on it

**Comment by laytan** [comment]
I couldn't really reproduce this either, but I did run your code with a debug build of llvm 14 (which has a bunch of assertions) and fixed a few here: 3b22c0854c0a64ae27ca43f59b5f6cdffcae56ab which makes it run without hitting any now. This happened to be a fix regarding `#soa` too, so maybe that was it?

**Comment by LeoBlute** [comment]
> I couldn't really reproduce this either, but I did run your code with a debug build of llvm 14 (which has a bunch of assertions) and fixed a few here: [3b22c08](https://github.com/odin-lang/Odin/commit/3b22c0854c0a64ae27ca43f59b5f6cdffcae56ab) which makes it run without hitting any now. This happened to be a fix regarding `#soa` too, so maybe that was it?

Yes, it's fixed
